# knitr document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - to date

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains. What we observed in earlier reports was that these genes manage to escape repression of their promoter by locally detaching from the lamina. Maybe there is a mechanism behind this where multiple factors can stimulate this behavior and the combination factors cooperating at a specific promoter dictate how much the promoter is extruded out of the lamina. One of such factors could be a CpG island. In almost all escapers CpG islands are found, but often these CpG islands have relatively low amount of CG's, short size and low CG O/E values. Maybe there if these CpG's were stronger, the HMM used to call the LAD states would detect the bigger extrusion from the lamina which would lead to a small iLAD created by this single promoter. Another posibility is that CTCF in fact plays a role in determining the size of the extrusion and when CTCF binding sites are further apart HMM calls an iLAD instead of a small dip in a LAD. 

## Description of Data.

gencode.sure.160617.rda: 
    file from Joris, received 17 June 2016. Promoter positions in this file are from Gencode. Contains SuRE, gro-cap and cage expression data as well as the number of tissues in which each promoter is expressed.


## libraries, paths and data prep

```{r, fig.width=10, fig.height=10}
library(reshape2)
library(rtracklayer)
library(DESeq2)
library(ggplot2)
library(gridExtra)
library(plyr)
library(preprocessCore)
library(scales)
library(gage)
library(fgsea)
library(CGtools)
library(biomaRt)
library(Matrix)
library(grid)
load('../raw_data/biomart.rdata')

## There was one promoter that was wrongly annotated
bm_p[bm_p$ensembl_transcript_id=='ENST00000357491','ensembl_gene_id' ] = 'ENSG00000196350'

```


# MAIN

### SuRE data
Previously, Bas defined three different groups of promoters in LADs. The "escaper", "repressed" and "inactive" group. The "escaper" group contains promoters that are expressed in their endogenous LAD environment. The "repressed" group is inactive in the endogenous setting, but they do show autonomous promoter activity in SuRE data. The 'inactive' group shows neither endogenous expression nor autonomous promoter activity in SuRE data.

```{r, fig.width=10, fig.height=8}
load("../raw_data/gencode.sure.160617.rda")
Prom<-gencode.sure.160617; rm(gencode.sure.160617) #simpler name
#first re-calculate pseudocounts without jitter
P<-Prom[,c(1:8,23,20,26,28, 27)] #SuRE, GRO-cap, CAGE and LAD columns only
names(P)[9:13]<-c("SuRE", "GROcap", "CAGE", "LAD", 'tissues_expressed')


## for promoters and gene expression let's convert promoter transcript id's to gene id's
P$ensembl_transcript_id = do.call(rbind, strsplit(P$name, split='[.]'))[,1]

nrow(P) #orriginal number of rows
bm_match = match(P$ensembl_transcript_id, bm_p$ensembl_transcript_id)
P<-merge(P, bm_p, by="ensembl_transcript_id", all.x=TRUE)
nrow(P) #some double rows were introduced

P = P[match(Prom$name, P$name), ]

length(unique(P$ensembl_gene_id)) #number of unique genes

table(P[,c('strand.x','strand.y')]) #almost all strand listings are consistent

P<-P[, colnames(P)!='strand.y']
colnames(P)[colnames(P)=='strand.x'] = "strand"

## to be used by CGtools as the complete set of TSS's
peaks = data.frame(seqname=P$chr,
                   start=P$tss,
                   end=P$tss,
                   strand=P$strand)


Pseud<-min(P$SuRE[P$SuRE>0], na.rm=TRUE)/2
P$SuRE<-P$SuRE+Pseud
P$SuRE<-log10(P$SuRE)
PseudGro<-min(P$GROcap[P$GROcap>0], na.rm=TRUE)/2
P$GROcap<-P$GROcap+PseudGro
P$GROcap<-log10(P$GROcap)
PseudCage<-min(P$CAGE[P$CAGE>0], na.rm=TRUE)/2
P$CAGE<-P$CAGE+PseudCage
P$CAGE<-log10(P$CAGE)

#then calculate running mean for iLAD promoters:
P<-P[order(P$SuRE,sample(c(1:nrow(P)))),] #sort by SuRE and then random for ties
n<-60 #number of windows
w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
s<-round(seq(from=w/2+0.0001, to=nrow(P)-w/2, length.out=n))
RM<-data.frame(SuRE.low=rep(NA,n), SuRE.mean=rep(NA,n), SuRE.hi=rep(NA,n), GROcap.lad=rep(NA,n), GROcap.ilad=rep(NA,n))
RM$SuRE.low=P$SuRE[s-floor(w/2)]
for(i in 1:n){RM$SuRE.mean[i]=mean(P$SuRE[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
RM$SuRE.hi=P$SuRE[s+floor(w/2)]
for(i in 1:n)
  {t<-P[(s[i]-floor(w/2)):(s[i]+floor(w/2)),]
   RM$GROcap.lad[i]<-mean(t$GROcap[t$LAD==1], na.rm=TRUE)
   RM$GROcap.ilad[i]<-mean(t$GROcap[t$LAD==0], na.rm=TRUE)
  }


#add first datapoint (SuRE equals pseudocount)
RM1<-RM[0,] #empty df
RM1[1,]<-c(rep(log10(Pseud),3), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==1]), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==0]))
RM<-rbind(RM1, RM)
rm(RM1)

#finally calculate LRS for all promoters:
P$LRS<- P$GROcap - approx(x=RM$SuRE.mean, y=RM$GROcap.ilad, xout=P$SuRE, rule=2)$y
#so the more negative the score, the more 'repressed' is the promoter by its chromatin/LAD context


#(arbitrary) cutoffs to define three groups of promoters:
INACT<- P$SuRE< -0.3 & P$LAD==1 & P$GROcap< -2 #inactive
NREP<- P$SuRE> 0 & P$LRS> -0.5 & P$LAD==1 & P$GROcap> -2 #not repressed
REP<- P$SuRE> 0.3 & P$LRS< -1 & P$LAD==1  & P$GROcap< -2 #repressed
Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
names(Pcnts)<-c("repressed", "escaper", "inactive")
BND <- P$LAD==1 & !INACT & !NREP & !REP

#add class annotation column to P:
P$class<-NA
P$class[P$LAD==0]<-"iLAD"
P$class[INACT]<-"inactive"
P$class[NREP]<-"escaper"
P$class[REP]<-"repressed"
P$class[BND] <- "boundary"
P$class = factor(P$class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary'))

COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")

class_names = paste0(levels(P$class), '; n=',table(P$class))
names(class_names) = levels(P$class)
P$class_n_prom = P$class
levels(P$class_n_prom) = class_names
COL_class_n_prom = COL_class[names(class_names)]
names(COL_class_n_prom) = class_names

lad_names = c(LAD=paste0('LAD; n=', table(P$LAD)['1']),
              iLAD=paste0('LAD; n=', table(P$LAD)['0']))
P$lad_n_prom = factor(ifelse(P$LAD==1, lad_names['LAD'], lad_names['iLAD']))
COL_lad_n = COL_lad
names(COL_lad_n) = lad_names


RM_melt = melt(RM, measure.vars=c('GROcap.ilad', 'GROcap.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='GROcap.lad', lad_names['LAD'], lad_names['iLAD'])
ggplot(P, aes(x=SuRE, y=GROcap, color=lad_n_prom)) +
    geom_point(data=P[P$LAD==0, ], size=0.5, alpha=0.05) + 
    geom_point(data=P[P$LAD==1, ], size=0.5, alpha=0.2) + 
    theme_bw() +
    geom_line(data=RM_melt, aes(x=SuRE.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_n)
p_classes = P[which(P$class %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE, y=GROcap)) + 
    geom_point(size=0.1,color=COLi) + 
    geom_point(data=p_classes, aes(color=class_n_prom), size=0.6) + 
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme_bw() +
    geom_line(data=RM, aes(x=SuRE.mean, y=GROcap.ilad), color=COL_lad['iLAD']) +
    theme(legend.title=element_blank()) +
    scale_colour_manual(values=COL_class_n_prom) 

```


## lonely genes.
```{r, fig.width=15, fig.height=15}
gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                              start=P$txStart,
                                              end=P$txEnd,
                                              strand=P$strand,
                                              tss=P$tss,
                                              name=P$name,
                                              gene_id = P$ensembl_gene_id),
                                              keep.extra.columns=TRUE)
tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss,
                         gene_gr$tss)
F_window = 22000

lad_gr = import.bed('../../../data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')
o = findOverlaps(lad_gr, tss_gr)
o_frame = data.frame(o)
o_frame$gene_id = tss_gr$gene_id[to(o)]

unique_frame = unique(o_frame[,c('queryHits', 'gene_id')])
unique_frame = unique_frame[!is.na(unique_frame$gene_id), ]
unique_frame = unique_frame[which(unique_frame$gene_id!='NA'), ]

lad_table = table(unique_frame$queryHits)

lonely_gr = lad_gr[as.numeric(names(lad_table)[lad_table==1])]

lonely_ilad_gr = lonely_gr[lonely_gr$name=='interLAD']
o_lonely = findOverlaps(lonely_ilad_gr, tss_gr)
lonely_p = P[to(o_lonely), ]

ggplot(P, aes(x=SuRE, y=GROcap, color=lad_n_prom)) +
    geom_point(data=P[P$LAD==0, ], size=0.5, alpha=0.05, color=COL_lad['iLAD']) + 
    geom_point(data=P[P$LAD==1, ], size=0.5, alpha=0.2, color=COL_lad['LAD']) + 
    geom_point(data=lonely_p, size=1, alpha=1, color='green') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=SuRE.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') 
```



```
nice -19 bedtools coverage -d \
                  -a ~/data/tracks/hg19/K562_CTCF_ChIP_narrow_ENCFF002CWL.bed.gz \
                  -b raw_data/prom_window.bed | \
                  awk '{if ($8 > 0){ print $0 }}' | \
                  awk '{
                      if (NR==1){
                          start=$1"\t"$2"\t"$3"\t"$4
                          valLine = $8
                          iLine = $7
                      }
                      else if (start==$1"\t"$2"\t"$3"\t"$4){
                          valLine = valLine"\t"$8
                          iLine = iLine"\t"$7
                      } else {
                          print start"\t"iLine;
                          print start"\t"valLine;
                          start=$1"\t"$2"\t"$3"\t"$4;
                          valLine = $8;
                          iLine = $7;
                      }
                  }END{
                      print start"\t"iLine;
                      print start"\t"valLine;
                  }' | gzip -c > \
                  raw_data/prom_window_CTCF.txt.gz &

``` 

```{r, fig.width=15, fig.height=15}
ctcf_table = read.table('../../../data/tracks/hg19/K562_CTCF_ChIP_narrow_ENCFF002CWL.bed.gz')
colnames(ctcf_table) = c('seqnames', 'start', 'end', 'name', 'score', 'strand', 'signalValue', 'pValue', 'qValue', 'point_source')
ctcf_gr = GRanges(seqnames=ctcf_table$seqnames, IRanges(ctcf_table$start+ctcf_table$point_source, ctcf_table$start+ctcf_table$point_source+1))
o = findOverlaps(tss_gr, ctcf_gr, maxgap=F_window, ignore.strand=T)

o_frame = data.frame(o)
o_frame$j = start(ctcf_gr[to(o)]) - start(tss_gr[from(o)]) + F_window
o_frame$j = ceiling(o_frame$j/100)


ctcf_matrix = matrix(nrow=nrow(P), ncol=440)

ctcf_matrix[as.matrix(o_frame[,c(1,3)])] = 1


plot_list = list()
for(i in c('iLAD', 'escaper', 'repressed', 'inactive')) #for the three gene classes
   {
    s = which(P[, 'class']==i)
    strand = P[,'strand'][s]
    this_window = rbind(ctcf_matrix[s[strand=='+'], ],
                        ctcf_matrix[s[strand=='-'],440:1])

    mean_list = lapply(seq(1,440,4), function(x){
                           c(x*100-50, sum(this_window[,x:(x+3)], na.rm=T)/length(s)/4)
                       })
    mean = do.call(rbind.data.frame, mean_list)
    colnames(mean) = c('x', 'y')
    #plot:
    plot_list[[i]] = ggplot(mean, aes(x=x-F_window, y=y)) + geom_line(col=COL_class[i]) + 
                        ggtitle(paste(i, '; n=', length(s))) + ylim(0,1) +
                        geom_vline(xintercept=0, lty='dotted') + 
                        theme_bw() +
                        theme(axis.title=element_blank()) +
                        scale_x_continuous(breaks=seq(-F_window, F_window, F_window))
}
do.call(grid.arrange, c(plot_list, nrow=1))
```