# knitr document van Steensel lab

# Promoters looping out of LADs
## Christ Leemans, 03-11-2016 - to date

## Introduction
We would like to explore the effect of LADs on gene expression. When comparing Joris' SuRE data which explorers promoter activity outside of the dna-context (like lamina assosiation) and GroCAP data which explorers promoter activity in the context, one can define 3 groups of promoter in LADs: repressed, inactive and non repressed. The inactive promoters show no activity in both groCAP and SuRE data, the repressed show activity only in SuRE data, the non repressed group show activity in groCAP data as well as SuRE data. 
We would like to see what defines these three groups.

This is a continuation of 'cl20161104_SuRE_in_differential_LADs.Rmd'


## libraries, paths and data prep

```{r, fig.width=10, fig.height=15}
library(reshape2)
library(rtracklayer)
library(DESeq2)
library(ggplot2)
library(gridExtra)
library(plyr)
library(preprocessCore)
library(scales)
load('../raw_data/biomart.rdata')

```

### gene expression

```{r, fig.width=10, fig.height=10}
file_list = list.files('../raw_data/expression', pattern='ReadsPerGene', full.names=T)

sample_id_list = lapply(file_list, function(x){
                        file = strsplit(x,'/')[[1]][4]
                        return(gsub('ReadsPerGene.out.tab', '', file))
                        })
count_list = lapply(file_list, function(x){
                count_table = read.table(x, stringsAsFactors=F)
                colnames(count_table) = c('name', 'unstranded', 'sense', 'antisense')
                return(count_table)
            })
names(count_list) = unlist(sample_id_list)


## since there are different machines, they might have different GC biases
## for GC bias we have to use hg19 genome
gff_ranges = import.gff('../../../data/GRCh37/Homo_sapiens.GRCh37.85.gff3')
## let's focus on genes located on actual chromosomes
gff_ranges = gff_ranges[seqnames(gff_ranges) %in% seqlevels(gff_ranges)[1:25]]
seqlevels(gff_ranges) = seqlevels(gff_ranges)[1:25]

ens_gene_ranges = gff_ranges[gff_ranges$gene_id %in% count_list[[1]][,1]]



colData = DataFrame(cell_type=c(rep('ESC', 2), rep('RPE',3), rep("HT1080", 3),
                                rep("Jurkat", 2), rep('K562', 2), rep('SupT1', 3)),
                    row.names=c('H7_hESC_rep1', 'H7_hESC_rep2',
                                'hfRPESeq012611_GSM898967', 'hfRPESeq090910_GSM898966', 
                                'hfRPESeq072611_GSM1099813',
                                'HT1080_ctrl_rep1_GSM1483942',
                                'HT1080_ctrl_rep2_GSM1483943',
                                'HT1080_ctrl_rep3_GSM1483944', 'J4_ctrl_rep1',
                                'J4_ctrl_rep2', 'K562_rep1', 'K562_rep2',
                                'SupT1_Mock_12hr_rep1', 'SupT1_Mock_12hr_rep2',
                                'SupT1_Mock_12hr_rep3'))

unstranded_list = lapply(count_list,function(x){
        count = x[5:nrow(x),2]
        names(count) = x[5:nrow(x), 1]
        return(count)
    })
count_table = do.call(cbind, unstranded_list)
colnames(count_table) = names(count_list)
count_table = count_table[,-grep('KBM7', names(count_list))]

## without quantile normalization
exp = SummarizedExperiment(assays=list(counts=count_table[, rownames(colData)]), rowRanges=ens_gene_ranges, colData=colData)
dds_direct = DESeqDataSet(exp, design = ~ cell_type)
dds_direct = DESeq(dds_direct)
fpm_direct = fpm(dds_direct)
fpm_direct = fpm_direct + min(fpm_direct[fpm_direct!=0])/2
ggplot(melt(fpm_direct), aes(x=log10(value), color=Var2)) + geom_density(adjust=1/2) + ggtitle('fpm values')


```


### SuRE data

```{r, fig.width=10, fig.height=15}
load("../raw_data/gencode.sure.160617.rda")
Prom<-gencode.sure.160617; rm(gencode.sure.160617) #simpler name
#first re-calculate pseudocounts without jitter
P<-Prom[,c(1:8,23,20,26,28, 27)] #SuRE, GRO-cap, CAGE and LAD columns only
names(P)[9:13]<-c("SuRE", "GROcap", "CAGE", "LAD", 'tissues_expressed')
Pseud<-min(P$SuRE[P$SuRE>0], na.rm=TRUE)/2
P$SuRE<-P$SuRE+Pseud
P$SuRE<-log10(P$SuRE)
PseudGro<-min(P$GROcap[P$GROcap>0], na.rm=TRUE)/2
P$GROcap<-P$GROcap+PseudGro
P$GROcap<-log10(P$GROcap)

#then calculate running mean for iLAD promoters:
P<-P[order(P$SuRE,sample(c(1:nrow(P)))),] #sort by SuRE and then random for ties
n<-60 #number of windows
w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
s<-round(seq(from=w/2+0.0001, to=nrow(P)-w/2, length.out=n))
RM<-data.frame(SuRE.low=rep(NA,n), SuRE.mean=rep(NA,n), SuRE.hi=rep(NA,n), GROcap.lad=rep(NA,n), GROcap.ilad=rep(NA,n))
RM$SuRE.low=P$SuRE[s-floor(w/2)]
for(i in 1:n){RM$SuRE.mean[i]=mean(P$SuRE[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
RM$SuRE.hi=P$SuRE[s+floor(w/2)]
for(i in 1:n)
  {t<-P[(s[i]-floor(w/2)):(s[i]+floor(w/2)),]
   RM$GROcap.lad[i]<-mean(t$GROcap[t$LAD==1], na.rm=TRUE)
   RM$GROcap.ilad[i]<-mean(t$GROcap[t$LAD==0], na.rm=TRUE)
  }
#add first datapoint (SuRE equals pseudocount)
RM1<-RM[0,] #empty df
RM1[1,]<-c(rep(log10(Pseud),3), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==1]), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==0]))
RM<-rbind(RM1, RM)
rm(RM1)

#finally calculate LRS for all promoters:
P$LRS<- P$GROcap - approx(x=RM$SuRE.mean, y=RM$GROcap.ilad, xout=P$SuRE, rule=2)$y
#so the more negative the score, the more 'repressed' is the promoter by its chromatin/LAD context




#(arbitrary) cutoffs to define three groups of promoters:
# LRS > 0 vs LRS < -1
ggplot(P[P$LAD==1,],aes(x=LRS)) + geom_density() + geom_vline(xintercept=c(0, -1))

# SuRE > 0.3 vs SuRE < -0.3 vs SuRE > 0
ggplot(P[P$LAD==1,],aes(x=SuRE)) + geom_density() + geom_vline(xintercept=c(0, -0.3, 0.3))


INACT<- P$SuRE< -0.3 & P$LAD==1 & P$GROcap< -2 #inactive
NREP<- P$SuRE>0 & P$LRS> 0 & P$LAD==1 #not repressed
REP<- P$SuRE>0.3 & P$LRS< -1 & P$LAD==1  & P$GROcap< -2 #repressed
Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
names(Pcnts)<-c("repressed", "not_repressed", "inactive")


#add class annotation column to P:
P$class<-NA
P$class[P$LAD==0]<-"iLAD"
P$class[INACT]<-"inactive"
P$class[NREP]<-"not_repressed"
P$class[REP]<-"repressed"

COLi<-"#00BBFF11" #dot color for iLAD promoters
#color vector for plotting:
COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "not_repressed", "inactive")
p_classes = cbind.data.frame(P, color=COL[P$class])
p_classes = p_classes[P$class %in% c('inactive', 'not_repressed', 'repressed'),]
ggplot(P, aes(x=SuRE, y=GROcap)) + 
    geom_point(size=0.1,color=COLi) + 
    geom_point(data=p_classes, aes(color=class), size=0.6)

```


### LAD-data

```{r, fig.width=10, fig.height=15}
#DamID data:
load("../raw_data/CdG140714humanLmnb1wK562.rData") #Lamin B1 DamID data compiled by Carolyn de Graaf
## make column names more descriptive of sample
colnames(allHumanAvHg19)[5:16] = 
    colnames(allHumanStateHg19)[5:16] = c('Tig3', 'ESC', 'HT1080', 'Jurkat', 'SupT1',
                                          'T2N', 'T1DS', 'LS174T', 'RPE', 'Tig3Guelen',
                                          'KBM7', 'K562')

## remove T1DS and T2N (fibroblasts of down syndrom twin) and Tig3Guelen.
## because T1DS and T2N have weird correlation with gene expression and Tig3Guelen
## is from a different platform
cell_selection = !colnames(allHumanAvHg19) %in% c('T2N', 'T1DS', 'Tig3Guelen')

lad_avg_probe = allHumanAvHg19[, cell_selection]
lad_hmm = allHumanStateHg19[, cell_selection]

```

### CpG island data

I previously downloaded CpG island data from UCSC:

orriginal file: cpgIslandExtUnmasked.txt.gz

source:
  http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/cpgIslandExtUnmasked.txt.gz

command:
  wget --timestamping http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/cpgIslandExtUnmasked.txt.gz   
  gunzip -c cpgIslandExtUnmasked.txt.gz â”‚| awk '{print $2"\t"$3"\t"$4"\t"$5$6}' > cpgIslandExtUnmasked_140601.bed

last modified: 01-06-14

doi:10.1016/0022-2836(87)90689-9



```{r, fig.width=10, fig.height=15}
cpg_ranges = import.bed('../raw_data/cpgIslandExtUnmasked_140601.bed')

dnase_file_list = list(Jurkat_rep1 = 'Jurkat_E61_rep1_ENCSR000EOS_hotspots_broadPeak.bed',
                       Jurkat_rep2 = 'Jurkat_E61_rep2_ENCSR000EOS_hotspots_broadPeak.bed',
                       K562_rep1 = 'K562_rep1_ENCSR000EPC_hotspots_broadPeak.bed',
                       K562_rep2 = 'K562_rep2_ENCSR000EPC_hotspots_broadPeak.bed',
                       KBM7_rep1 = 'KBM7_rep1_ENCSR426IEA_hotspots_broadPeak.bed',
                       KBM7_rep2 = 'KBM7_rep2_ENCSR426IEA_hotspots_broadPeak.bed')

dnase_list = lapply(names(dnase_file_list), function(x){
                        tab = read.table(paste0('../raw_data/DNase-seq/',
                                                dnase_file_list[[x]]),
                                         stringsAsFactors = F)
                        tab = tab[,c(1:3,9)]
                        colnames(tab) = c('seqnames', 'start', 'end', 'score')
                        return(makeGRangesFromDataFrame(tab, keep.extra.columns=TRUE))
                    })
names(dnase_list) = names(dnase_file_list)
```

### genomic ranges and links between data types.

```{r, fig.width=10, fig.height=15}
## for promoters and gene expression let's convert promoter transcript id's to gene id's
P$ensembl_transcript_id = do.call(rbind, strsplit(P$name, split='[.]'))[,1]

P<-merge(P, bm_p, by="ensembl_transcript_id", all.x=TRUE)
nrow(P) #number of unique promoters

length(unique(P$ensembl_gene_id)) #number of unique genes

table(P[,c('strand.x','strand.y')]) #almost all strand listings are consistent

P<-P[, colnames(P)!='strand.y']
colnames(P)[colnames(P)=='strand.x'] = "strand"

table(P$ensembl_gene_id %in% row.names(fpm_direct))
table(row.names(fpm_direct) %in% P$ensembl_gene_id)

#genes as Granges object
gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                              start=P$txStart,
                                              end=P$txEnd,
                                              strand=P$strand,
                                              tss=P$tss,
                                              ensembl_gene_id=P$ensembl_gene_id,
                                              ensembl_transcript_id=P$ensembl_transcript_id),
                                              keep.extra.columns=TRUE) 

lad_gr <- makeGRangesFromDataFrame(data.frame(seqnames = lad_hmm$seqname,
                                              start = lad_hmm$start,
                                              end = lad_hmm$end,
                                              row.names=rownames(lad_hmm)))

tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss - 1 * as.numeric(strand(gene_gr)=='-'),
                         gene_gr$tss + 1 * as.numeric(strand(gene_gr)=='+'))

tss_gr$cpg_distance = mcols(distanceToNearest(tss_gr, cpg_ranges))$distance

dnase_prom = data.frame(Jurkat=NA, K562=NA, KBM7=NA)
for (cell_type in c('Jurkat', 'KBM7', 'K562')){
    distance = data.frame(row.names=tss_gr$ensembl_transcript_id)
    for (rep in c('rep1', 'rep2')){
        name = paste(cell_type, rep, sep='_')
        dnase_ranges = dnase_list[[name]]
        distance[,name] = mcols(distanceToNearest(tss_gr, dnase_ranges))$distance
    }

    dnase_prom[which(rowSums(distance<1000)==2), cell_type] = 'dnase'
    dnase_prom[which(rowSums(distance>=3000)==2), cell_type] = 'non_dnase'
}
P$Jurkat_dnase = dnase_prom[,'Jurkat']
P$KBM7_dnase = dnase_prom[,'KBM7']
P$K562_dnase = dnase_prom[,'K562']


P$cpg_class = NA
cpg_prom = tss_gr[which(tss_gr$cpg_distance < 1000)]$ensembl_transcript_id

non_cpg_prom = tss_gr[which(tss_gr$cpg_distance >= 1000)]$ensembl_transcript_id

P$cpg_class[P$ensembl_transcript_id %in% cpg_prom] = 'CpG'
P$cpg_class[P$ensembl_transcript_id %in% non_cpg_prom] = 'non_CpG'


lad_overlap = findOverlaps(gene_gr, lad_gr)

gene_avg_lad = aggregate(subjectHits(lad_overlap), 
                         list(queryHits(lad_overlap)), function(x){
                            colMeans(lad_avg_probe[x,c('Jurkat','KBM7','K562')])
                         })

gene_avg_hmm_lad = aggregate(subjectHits(lad_overlap), 
                         list(queryHits(lad_overlap)), function(x){
                            colMeans(lad_hmm[x,c('Jurkat','KBM7','K562')])
                         })

P$K562_avg_lad = NaN
P$KBM7_avg_lad = NaN
P$Jurkat_avg_lad = NaN
P[gene_avg_lad[,1], 'K562_avg_lad'] = gene_avg_lad$x[,'K562']
P[gene_avg_lad[,1], 'Jurkat_avg_lad'] = gene_avg_lad$x[,'Jurkat']
P[gene_avg_lad[,1], 'KBM7_avg_lad'] = gene_avg_lad$x[,'KBM7']

P$K562_hmm_lad = NaN
P$KBM7_hmm_lad = NaN
P$Jurkat_hmm_lad = NaN

P[gene_avg_hmm_lad[,1], 'K562_hmm_lad'] = gene_avg_hmm_lad$x[,'K562']
P[gene_avg_hmm_lad[,1], 'KBM7_hmm_lad'] = gene_avg_hmm_lad$x[,'KBM7']
P[gene_avg_hmm_lad[,1], 'Jurkat_hmm_lad'] = gene_avg_hmm_lad$x[,'Jurkat']



## full region around TSS
F_full = 22000
full_region_gr = gene_gr
full_start = full_region_gr$tss - F_full
full_start = ifelse(full_start<1,1,full_start)
full_end = full_region_gr$tss + F_full
ranges(full_region_gr) = IRanges(full_start, full_end)

```

```{r, fig.width=10, fig.height=15}
F=22000

Egr.p = gene_gr[width(gene_gr)>F]
p_match = match(Egr.p$ensembl_gene_id, P$ensembl_gene_id)
P_start = P[p_match, 'tss'] - F
P_start = ifelse(P_start<1,1,P_start)
P_end = P[p_match, 'tss'] + F
ranges(Egr.p) = IRanges(P_start, P_end)

h<-findOverlaps(Egr.p, lad_gr)
oENSG<-Egr.p[from(h)]$ensembl_gene_id
oPOS<-ifelse(strand(Egr.p[from(h)])=='+',
             (start(lad_gr[to(h)])+end(lad_gr[to(h)]))/2 - start(Egr.p[from(h)]) -F,
             end(Egr.p[from(h)]) - (start(lad_gr[to(h)])+end(lad_gr[to(h)]))/2 -F)

         #coordinates of all overlapping probes relative to the gene starts
oVAL<- lad_avg_probe[names(lad_gr[to(h)]),]


#plot:
par(mfrow=c(1,6))
for(celltype in c('Jurkat', 'KBM7', 'K562'))
   {for (dnase_class in c('dnase', 'non_dnase')){
        dnase_name = paste0(cell_type, '_dnase')
        hmm_name = paste0(cell_type, '_hmm_lad')
        s<-unique(P$ensembl_gene_id[which(P[,hmm_name]==2 & P[,dnase_name]==dnase_class)])
        s<-s[!is.na(s)]
        w<-oENSG %in% s #which rows in oENSG correspond to genes in s
        subPOS<-oPOS[w]
        subVAL<-oVAL[w, celltype]
        o<-order(subPOS) #need to order all values by distance to gene start for running median plot:
        subPOS<-subPOS[o]
        subVAL<-subVAL[o]
        #determine runmed k:
        wsize<-floor(length(subPOS)/25) #4% of all datapoints in the set
        if(!wsize %% 2) {wsize<-wsize+1} #must be odd
        #plot:
        plot(subPOS, subVAL, pch=".", col="#00000033", xlim=c(-F,F)*0.9, ylim=c(-0.5,1),
             main=paste(celltype, dnase_class, length(s)), xlab="position relative to gene start (bp)", ylab="DamID logratio")
        lines(subPOS, runmed(subVAL, k=wsize), col=COL[i], lwd=2)
        abline(v=0, lty="dotted")
     }
    
 }
```

**conclusion:**
DNase signal close to the promoter of the LAD-gene often means a dip in lamina association. But probably this mainly correlates with CpG islands.


## DamID signal split by DNase and CpG 

```{r, fig.width=10, fig.height=15}
ggplot(P, aes(x=GROcap)) + geom_density() + geom_vline(xintercept=-3) + geom_vline(xintercept=-1)


par(mfrow=c(1,4))
for(celltype in c('KBM7', 'Jurkat', 'K562'))
 {for (dnase_class in c('dnase', 'non_dnase')) 
   {for (cpg_class in c('CpG', 'non_CpG')){
        dnase_name = paste0(cell_type, '_dnase')
        hmm_name = paste0(cell_type, '_hmm_lad')
        selection = which(P[,hmm_name]==2 & P[,dnase_name]==dnase_class & P$cpg_class==cpg_class)
        s<-unique(P$ensembl_gene_id[selection])
        s<-s[!is.na(s)]
        w<-oENSG %in% s #which rows in oENSG correspond to genes in s
        subPOS<-oPOS[w]
        subVAL<-oVAL[w, celltype]
        o<-order(subPOS) #need to order all values by distance to gene start for running median plot:
        subPOS<-subPOS[o]
        subVAL<-subVAL[o]
        #determine runmed k:
        wsize<-floor(length(subPOS)/25) #4% of all datapoints in the set
        if(!wsize %% 2) {wsize<-wsize+1} #must be odd
        title = paste(celltype, dnase_class, '\n', length(s), cpg_class, '\ncomplete gene in LAD hmm')
        #plot:
        plot(subPOS, subVAL, pch=".", col="#00000033", xlim=c(-F,F)*0.9, ylim=c(-0.5,1),
             main=title, xlab="position relative to gene start (bp)", ylab="DamID logratio")
        lines(subPOS, runmed(subVAL, k=wsize), col=COL[i], lwd=2)
        abline(v=0, lty="dotted")
     }
    
   }
 }


oLAD<- lad_hmm[names(lad_gr[to(h)]),]
par(mfrow=c(1,4))
for(celltype in unique(colData$cell_type))
 {for(active in c(TRUE, FALSE)) 
   {for (cpg_class in c('CpG', 'non_CpG')){
        if (active){
            selection = which(P$GROcap > -1 & P$cpg_class==cpg_class)
        } else {
            selection = which(P$GROcap < -3 & P$cpg_class==cpg_class)
        }
        s<-unique(P$ensembl_gene_id[selection])
        s<-s[!is.na(s)]
        subLAD<-unlist(lapply(s, function(x){
            this_hmm = table(oLAD[oENSG==x,celltype])
            if (length(this_hmm)==2){
                return(this_hmm['2']/sum(this_hmm))
            } else if ('2'%in%names(this_hmm)){
                return(1)
            } else {
                return(0)
            }}))
        s <- s[subLAD>0.8]
        w<-oENSG %in% s #which rows in oENSG correspond to genes in s
        subPOS<-oPOS[w]
        subVAL<-oVAL[w, celltype]

        o<-order(subPOS) #need to order all values by distance to gene start for running median plot:
        subPOS<-subPOS[o]
        subVAL<-subVAL[o]
        #determine runmed k:
        wsize<-floor(length(subPOS)/25) #4% of all datapoints in the set
        if(!wsize %% 2) {wsize<-wsize+1} #must be odd
        title = ifelse(active, paste(celltype, 'GROcap > -1\n', length(s), cpg_class,
                                     '\nLAD in cell type'),
                       paste(celltype, 'GROcap < -3\n', length(s), cpg_class, 
                             '\nLAD in cell type'))
        #plot:
        plot(subPOS, subVAL, pch=".", col="#00000033", xlim=c(-F,F)*0.9, ylim=c(-0.5,1),
             main=title, xlab="position relative to gene start (bp)", ylab="DamID logratio")
        lines(subPOS, runmed(subVAL, k=wsize), col=COL[i], lwd=2)
        abline(v=0, lty="dotted")
     }
    
   }
 }


```

**conclusion:**

It really seems like CpG island promoters escape lamina interaction, even if they are not active. But also in non-CpG promoters there seems to be a tendency of the promoter to escape the Lamina environment.